File: utils/rag.js
const fs = require("fs");
const path = require("path");

function simpleRetrieve(query, k = 3) {
  const kbPath = path.join(__dirname, "..", "knowledge_base", "facts.txt");
  if (!fs.existsSync(kbPath)) return [];

  const facts = fs.readFileSync(kbPath, "utf-8")
    .split("\n")
    .map(f => f.trim())
    .filter(f => f.length > 0);

  const matched = facts
    .filter(f => query.toLowerCase().split(" ").some(word => f.toLowerCase().includes(word)))
    .slice(0, k);

  return matched;
}

module.exports = { simpleRetrieve };

File: utils/rag.py
import os

KB_PATH = os.path.join("knowledge_base", "facts.txt")

def load_kb():
    if not os.path.exists(KB_PATH):
        return []
    with open(KB_PATH, "r", encoding="utf-8") as f:
        return [ln.strip() for ln in f if ln.strip()]

def simple_retrieve(query, k=3):
    facts = load_kb()
    if not facts:
        return []
    q_tokens = set(query.lower().split())
    def score(f):
        ft = set(f.lower().split())
        inter = len(q_tokens & ft)
        union = len(q_tokens | ft) or 1
        return inter / union
    return sorted(facts, key=score, reverse=True)[:k]
